node/src/reactor/participating.rs

```diff
@@ -664,75 +665,96 @@ impl reactor::Reactor for Reactor {
-                            let message = Message::FinalitySignature(Box::new(signature));
-                            effect_builder.broadcast_message(message).await
+                                .announce_upcoming_era_validators(
+                                    current_era_id,
+                                    step_effect_and_upcoming_era_validators.upcoming_era_validators,
+                                )
+                                .ignore(),
+                        );
+                    }
+
+                    let secret_key = our_secret_key.clone();
+                    let public_key = our_public_key.clone();
+                    let block_hash = *block.hash();
+                    effects.extend(
+                        async move {
+                            let validator_weights =
+                                match linear_chain::era_validator_weights_for_block(
+                                    block.header(),
+                                    effect_builder,
+                                )
+                                .await
+                                {
+                                    Ok((_era_id, weights)) => weights,
+                                    Err(error) => {
+                                        return fatal!(
+                                            effect_builder,
+                                            "couldn't get era validators for header: {}",
+                                            error
+                                        )
+                                        .await;
+                                    }
+                                };
+
+                            // We're responsible for signing the new block if we're in the provided list.
+                            if validator_weights.contains_key(&public_key) {
+                                let signature = FinalitySignature::new(
+                                    block_hash,
+                                    current_era_id,
+                                    &secret_key,
+                                    public_key.clone(),
+                                );
+
+                                effect_builder
+                                    .announce_created_finality_signature(signature.clone())
+                                    .await;
+                                // Allow a short period for peers to establish connections. This
+                                // delay can be removed once we move to a single reactor model.
+                                effect_builder
+                                    .set_timeout(DELAY_FOR_SIGNING_IMMEDIATE_SWITCH_BLOCK)
+                                    .await;
+                                let message = Message::FinalitySignature(Box::new(signature));
+                                effect_builder.broadcast_message(message).await;
+                            }
                         }
```

Comment:
This looks like something that could be extracted into a function like `sign_immediate_switch_block_if_needed` - only I'm not sure where it should live. We tend to avoid putting complex logic into `effect_builder`, so maybe `linear_chain`, like `era_validator_weights_for_block`?

Reply:
I believe a cleanup of effect_builder is already planned, when code like this will be moved out.  So while I agree, it's probably outwith the scope of this PR.

Reply:
The thing is, it's inlined in `Participating::new` right now - and that function is _huge_ as a result. I was only considering moving it to `effect_builder`, but decided against it. Do you think I should leave it as is, in `Participating::new`?